src/driver/alloy.rs

```diff
@@ -0,0 +1,659 @@
+        }
+
+        // Check if the bundle has any transactions
+        if self.bundle.txs.is_empty() {
+            return Err(trevm.errored(BundleError::BundleEmpty));
+        }
+
+        // Check if the state block number is valid (not 0, and not a tag)
+        if !self.bundle.state_block_number.is_number()
+            || self.bundle.state_block_number.as_number().unwrap_or(0) == 0
+        {
+            return Err(trevm.errored(BundleError::BlockNumberMismatch));
+        }
+
+        // Set the state block number this simulation was based on
+        self.response.state_block_number = self
+            .bundle
+            .state_block_number
+            .as_number()
+            .unwrap_or(trevm.inner().block().number.to::<u64>());
+
+        let bundle_filler = BundleBlockFiller::from(&self.bundle);
+
+        let run_result = trevm.try_with_block(&bundle_filler, |trevm| {
+            // We need to keep track of the state of the EVM as we run the transactions, so we can accumulate the results.
+            // Therefore we keep this mutable trevm instance, and set it to the new one after we're done simulating.
+            let mut trevm = trevm;
+
+            let txs = self
+                .bundle
+                .txs
+                .iter()
+                .map(|tx| TxEnvelope::decode_2718(&mut tx.chunk()))
+                .collect::<Result<Vec<_>, _>>();
+            let txs = match txs {
+                Ok(txs) => txs,
+                Err(e) => return Err(trevm.errored(BundleError::TransactionDecodingError(e))),
+            };
+
+            // Check that the bundle does not exceed the maximum gas limit for blob transactions
+            if txs
+                .iter()
+                .filter_map(|tx| tx.as_eip4844())
+                .map(|tx| tx.tx().tx().blob_gas())
+                .sum::<u64>()
+                > MAX_BLOB_GAS_PER_BLOCK
+            {
+                return Err(trevm.errored(BundleError::Eip4844BlobGasExceeded));
+            }
+
+            // Cache the pre simulation coinbase balance, so we can use it to calculate the coinbase diff after every tx simulated.
+            let initial_coinbase_balance =
+                match trevm.try_read_balance(trevm.inner().block().coinbase) {
+                    Ok(balance) => balance,
+                    Err(e) => {
+                        return Err(trevm.errored(BundleError::EVMError {
+                            inner: revm::primitives::EVMError::Database(e),
+                        }))
+                    }
+                };
+            let mut pre_sim_coinbase_balance = initial_coinbase_balance;
+            let post_sim_coinbase_balance = pre_sim_coinbase_balance;
+            let mut total_gas_fees = U256::ZERO;
+            let mut total_gas_used = 0;
+
+            let mut hash_bytes = Vec::with_capacity(32 * txs.len());
+
+            for tx in txs.iter() {
+                let run_result = trevm.run_tx(tx);
```

Comment:
I feel like the body here could be a function on self maybe? 🤔 it's big

Reply:
yup. abstracted in https://github.com/init4tech/trevm/pull/34/commits/299fb4b61db82ac66e7d675559227990258e56b4